src/evaluators/LogicalExpression.js

```diff
@@ -93,5 +93,12 @@ export default function(
   // return or throw completion
   if (completion instanceof AbruptCompletion) throw completion;
   invariant(completion instanceof NormalCompletion || completion instanceof Value || completion instanceof Reference);
+  if (lval instanceof Value && compl2 instanceof Value) {
+    // joinEffects does the right thing for the side effects of the second expression but for the result the join
+    // produces a conditional expressions of the form (a ? b : a) for a && b and (a ? a : b) for a || b
+    // Rather than look for this pattern everywhere, we override this behavior and replace the completion with
```

Comment:
"Rather than look for this pattern everywhere" - I would phrase that as a TODO: In some future, all such expressions should immediately simplify down to some normal form, and it should be impossible to construct values that are not in normal form.

Reply:
This is from the previous pull request. Adding a TODO like that at this point does not strike me as at all useful.